File: Montes-Guerrero-Daniel/filosofos-comelones/filosofos.c
/*
 * Alumno: Daniel Montes Guerrero
 * Compilar: gcc filosofos.c -lpthread
 */

#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<sys/ipc.h>
#include<sys/types.h>
#include<unistd.h>
#include<fcntl.h>
#include<sys/ipc.h>
#include<sys/sem.h>
#include<sys/wait.h>
#include<semaphore.h>
#include<pthread.h>
#include<math.h>

#define BUFF_SIZE 200

int max(int a, int b){
	return a > b ? a : b;
}

int num_fil;
sem_t *sem_eating, **sem_forks;

void* eat(void* arg){
	int id = *((int*)arg);
	while(1){
		sem_wait(sem_eating);
		sem_wait(sem_forks[id]);
		printf("Filosofo: %d obtiene tenedor %d\n", id, id);
		sem_wait(sem_forks[(id + 1) % num_fil]);
		printf("Filosofo: %d obtiene tenedor %d\n", id, (id + 1) % num_fil);
		printf("Filosofo: %d empieza a comer\n", id);
		sleep(2);
		printf("Filosofo: %d termina de comer\n", id);
		sem_post(sem_eating);
		sem_post(sem_forks[id]);
		sem_post(sem_forks[(id + 1) % num_fil]);
		sleep(1);
	}
}

int main(int argc, char** argv){
	if(argc < 2){
		printf("Uso: ./filosofos <num_filosofos>\n");
		return 1;
	}
	num_fil = atoi(argv[1]);
	int i = 0;
	sem_forks = (sem_t**)malloc(sizeof(sem_t*) * (num_fil));
	sem_unlink("sem_eating");
	sem_eating = sem_open("sem_eating", O_CREAT | O_EXCL, 0644, max(1, num_fil - 1));
	
	for(i = 0; i < num_fil; i++){
		char sem_name[BUFF_SIZE] = "sem_fork_";
		char num[BUFF_SIZE];
		snprintf(num, BUFF_SIZE, "%d", i);
		strcat(sem_name, num);
		printf("sem_name: %s\n", sem_name);
		sem_unlink(sem_name);
		sem_forks[i] = sem_open(sem_name, O_CREAT | O_EXCL, 0644, 1);
	}

	pthread_t* thread = (pthread_t*)malloc(sizeof(pthread_t) * num_fil);
	int* ids = (int*)malloc(sizeof(int) * num_fil);
	for(i = 0; i < num_fil; i++){
		ids[i] = i;
		if(pthread_create(&thread[i], NULL, eat, (void*)(&ids[i])) != 0){
			break;
		}
	}

	for(i = 0; i < num_fil; i++){
		pthread_join(thread[i], NULL);
	}

	return 0;
}

File: Montes-Guerrero-Daniel/filosofos-comelones/Filosofos.java
/*
 * Alumno: Daniel Montes Guerrero
 * Compilar: javac Filosofos.java
 */

import java.util.ArrayList;

class Controller {
    int num_eating;
    int num_fil;
    int limit;
    ArrayList<Integer> forks;

    Controller(int num_fil){
        num_eating = 0;
        forks = new ArrayList<Integer>(num_fil);        
        for(int i = 0; i < num_fil; i++){
            forks.add(0);
        }
        this.num_fil = num_fil;
        this.limit = Math.max(1, num_fil - 1);
    }

    public synchronized void wait_to_eat(){
        while(num_eating == limit){
            try{
                wait();
            }
            catch(Exception e){
                System.out.println(e);
                System.out.println("error en controlador");
            }
        }
        num_eating++;
        notifyAll();
    }
    public synchronized void wait_to_get_fork(int fork){
        while(forks.get(fork) == 1){
            try{
                wait();
            }
            catch(Exception e){
                System.out.println(e);
                System.out.println("error en controlador");
            }
        }
        forks.set(fork, 1);
        notifyAll();
    }
    public synchronized void release_eating(){
        num_eating--;
        notifyAll();
    }
    public synchronized void release_fork(int fork){
        forks.set(fork, 0);
        notifyAll();
    }
}

class Filosofo extends Thread {
    Controller controller;
    int id;
    int num_fil;

    Filosofo(Controller controller, int id, int num_fil){
        this.controller = controller;
        this.id = id;
        this.num_fil = num_fil;
    }

    public void run(){
        try{
            while (true) {
                controller.wait_to_eat();
                controller.wait_to_get_fork(id);
                System.out.println("Filosofo: " + id + " obtiene tenedor " + id);
                controller.wait_to_get_fork((id + 1) % num_fil);
                System.out.println("Filosofo: " + id + " obtiene tenedor " + (id + 1) % num_fil);
                System.out.println("Filosofo: " + id + " empieza a comer");
                Thread.sleep(2000);

                System.out.println("Filosofo: " + id + " termina de comer");
                controller.release_eating();
                controller.release_fork(id);
                controller.release_fork((id + 1) % num_fil);
                Thread.sleep(1000);
            }
        }
        catch(Exception e){
            System.out.println(e);
            System.out.println("Error en filosofo");
        }
    }
}

public class Filosofos {
    public static void main(String[] args){
        if (args.length < 1) {
            System.out.println("Uso: java Filosofos <num_filosofos>");
            return;
        }
        int num_fil = Integer.parseInt(args[0]);
        System.out.println("Filosofos: " + num_fil);
        var controller = new Controller(num_fil);
        var filosofos = new ArrayList<Filosofo>(num_fil);
        for (int i = 0; i < num_fil; i++) {
            filosofos.add(new Filosofo(controller, i, num_fil));
            filosofos.get(i).start();
        }
        try {
            for (int i = 0; i < num_fil; i++) {
                filosofos.get(i).join();
            }
        } catch (InterruptedException e) {

        }
    }
}